## Regression with Structured Data

Solve the California Housing Prices

a) Use 5-fold cross validation

In [11]:
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.datasets import fetch_california_housing

california_housing = fetch_california_housing(as_frame=True)

x = california_housing.data
y = california_housing.target

print(x)
print(y)

k = 5
kf = KFold(n_splits=k, shuffle=True, random_state=42)

       MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  
0        -122.23  
1

b) Compare Linear Regression and Decision Tree

In [12]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression


model_decision_tree = DecisionTreeRegressor()
model_linear_regression = LinearRegression()

c) Evaluate the MSE and MAE (not using Power Transform)

Observations:
- MSE:

Mean Squared Error is useful for problems that we want to penalize big errors/outliners.

  - MAE:

Mean Absolute Error is useful for problems that we want a fair and less affected by extremes average.

In [13]:
from sklearn.model_selection import cross_val_score
import numpy as np

scores_decision_tree_mse = cross_val_score(model_decision_tree, x, y, cv=kf, scoring='neg_mean_squared_error')
scores_decision_tree_mae = cross_val_score(model_decision_tree, x, y, cv=kf, scoring='neg_mean_absolute_error')
scores_linear_regression_mse = cross_val_score(model_linear_regression, x, y, cv=kf, scoring='neg_mean_squared_error')
scores_linear_regression_mae = cross_val_score(model_linear_regression, x, y, cv=kf, scoring='neg_mean_absolute_error')


avarage_decision_tree_mse = np.mean(scores_decision_tree_mse)
avarage_decision_tree_mae = np.mean(scores_decision_tree_mae)
avarage_linear_regression = np.mean(scores_linear_regression_mse)
avarage_linear_regression_mae = np.mean(scores_linear_regression_mae)

print("-Decision Tree:")
print("MSE: ", avarage_decision_tree_mse)
print("MAE: ", avarage_decision_tree_mae)

print("-Linear Regression:")
print("MSE: ", avarage_linear_regression)
print("MAE: ", avarage_linear_regression_mae)



-Decision Tree:
MSE:  -0.5117852964661822
MAE:  -0.46241199418604656
-Linear Regression:
MSE:  -0.5305724639238882
MAE:  -0.5316768254176778


d) Compare each model using power transform

In [14]:
from sklearn.preprocessing import PowerTransformer

pt = PowerTransformer(method='yeo-johnson')
x = pt.fit_transform(x)

print(x)
print(y)

k = 5
kf = KFold(n_splits=k, shuffle=True, random_state=42)

[[ 1.90328852e+00  9.75968374e-01  1.10639400e+00 ... -3.68763198e-01
   1.08903746e+00 -2.41473508e-15]
 [ 1.89708090e+00 -5.68892992e-01  7.12366596e-01 ... -1.18626485e+00
   1.08240369e+00 -2.41473508e-15]
 [ 1.60436139e+00  1.76101030e+00  1.67364461e+00 ...  5.05285191e-03
   1.07907856e+00 -2.41473508e-15]
 ...
 [-1.50942580e+00 -9.05850477e-01  4.92811385e-02 ... -7.64454822e-01
   1.54296057e+00 -1.58206781e-15]
 [-1.32391444e+00 -8.20371971e-01  1.37691863e-01 ... -1.15852734e+00
   1.54296057e+00 -1.66533454e-15]
 [-8.19529103e-01 -9.92245443e-01  8.46759977e-02 ... -2.71158648e-01
   1.52740291e+00 -1.60982339e-15]]
0        4.526
1        3.585
2        3.521
3        3.413
4        3.422
         ...  
20635    0.781
20636    0.771
20637    0.923
20638    0.847
20639    0.894
Name: MedHouseVal, Length: 20640, dtype: float64


Evaluate the MSE and MAE (using Power Transform)

In [15]:
from sklearn.model_selection import cross_val_score
import numpy as np

scores_decision_tree_mse = cross_val_score(model_decision_tree, x, y, cv=kf, scoring='neg_mean_squared_error')
scores_decision_tree_mae = cross_val_score(model_decision_tree, x, y, cv=kf, scoring='neg_mean_absolute_error')
scores_linear_regression_mse = cross_val_score(model_linear_regression, x, y, cv=kf, scoring='neg_mean_squared_error')
scores_linear_regression_mae = cross_val_score(model_linear_regression, x, y, cv=kf, scoring='neg_mean_absolute_error')

avarage_decision_tree_mse_pt = np.mean(scores_decision_tree_mse)
avarage_decision_tree_mae_pt = np.mean(scores_decision_tree_mae)
avarage_linear_regression_pt = np.mean(scores_linear_regression_mse)
avarage_linear_regression_mae_pt = np.mean(scores_linear_regression_mae)

print("-Decision Tree:")
print("MSE: ", avarage_decision_tree_mse_pt)
print("MAE: ", avarage_decision_tree_mae_pt)

print("-Linear Regression:")
print("MSE: ", avarage_linear_regression_pt)
print("MAE: ", avarage_linear_regression_mae_pt)


-Decision Tree:
MSE:  -0.5195009782754603
MAE:  -0.4609438716085272
-Linear Regression:
MSE:  -0.5691570624970801
MAE:  -0.5710045544688994


Evaluate the MSE and MAE (using Z-scale and Min-max Scale)

In [16]:
from sklearn.preprocessing import StandardScaler

scaler_z = StandardScaler()
x = scaler_z.fit_transform(x)

print(x)
print(y)

k = 5
kf = KFold(n_splits=k, shuffle=True, random_state=42)

[[ 1.90328852  0.97596837  1.106394   ... -0.3687632   1.08903747
  -1.30333547]
 [ 1.8970809  -0.56889299  0.7123666  ... -1.18626485  1.0824037
  -1.30333547]
 [ 1.60436139  1.7610103   1.67364461 ...  0.00505285  1.07907857
  -1.30333547]
 ...
 [-1.5094258  -0.90585048  0.04928114 ... -0.76445482  1.54296058
  -0.85459043]
 [-1.32391444 -0.82037197  0.13769186 ... -1.15852734  1.54296058
  -0.89946494]
 [-0.8195291  -0.99224544  0.084676   ... -0.27115865  1.52740292
  -0.8695486 ]]
0        4.526
1        3.585
2        3.521
3        3.413
4        3.422
         ...  
20635    0.781
20636    0.771
20637    0.923
20638    0.847
20639    0.894
Name: MedHouseVal, Length: 20640, dtype: float64


In [17]:
from sklearn.model_selection import cross_val_score
import numpy as np

scores_decision_tree_mse = cross_val_score(model_decision_tree, x, y, cv=kf, scoring='neg_mean_squared_error')
scores_decision_tree_mae = cross_val_score(model_decision_tree, x, y, cv=kf, scoring='neg_mean_absolute_error')
scores_linear_regression_mse = cross_val_score(model_linear_regression, x, y, cv=kf, scoring='neg_mean_squared_error')
scores_linear_regression_mae = cross_val_score(model_linear_regression, x, y, cv=kf, scoring='neg_mean_absolute_error')

avarage_decision_tree_mse_z = np.mean(scores_decision_tree_mse)
avarage_decision_tree_mae_z = np.mean(scores_decision_tree_mae)
avarage_linear_regression_z = np.mean(scores_linear_regression_mse)
avarage_linear_regression_mae_z = np.mean(scores_linear_regression_mae)

print("-Decision Tree:")
print("MSE: ", avarage_decision_tree_mse_z)
print("MAE: ", avarage_decision_tree_mae_z)

print("-Linear Regression:")
print("MSE: ", avarage_linear_regression_z)
print("MAE: ", avarage_linear_regression_mae_z)


-Decision Tree:
MSE:  -0.5185766439499079
MAE:  -0.461628316375969
-Linear Regression:
MSE:  -0.5077893876193731
MAE:  -0.5357830764989724


In [18]:
from sklearn.preprocessing import MinMaxScaler

scaler_minmax = MinMaxScaler()
x = scaler_minmax.fit_transform(x)

print(x)
print(y)

k = 5
kf = KFold(n_splits=k, shuffle=True, random_state=42)

[[0.81141396 0.81556942 0.61041951 ... 0.48914506 0.74535334 0.16524217]
 [0.81047236 0.4526338  0.58315641 ... 0.42325895 0.74365129 0.16524217]
 [0.76607152 1.         0.64966809 ... 0.51927256 0.74279816 0.16524217]
 ...
 [0.29376004 0.37347208 0.53727693 ... 0.45725451 0.86181786 0.25071225]
 [0.32189913 0.39355362 0.54339415 ... 0.42549444 0.86181786 0.24216524]
 [0.39840628 0.35317524 0.53972594 ... 0.49701145 0.85782618 0.24786325]]
0        4.526
1        3.585
2        3.521
3        3.413
4        3.422
         ...  
20635    0.781
20636    0.771
20637    0.923
20638    0.847
20639    0.894
Name: MedHouseVal, Length: 20640, dtype: float64


In [19]:
from sklearn.model_selection import cross_val_score
import numpy as np

scores_decision_tree_mse = cross_val_score(model_decision_tree, x, y, cv=kf, scoring='neg_mean_squared_error')
scores_decision_tree_mae = cross_val_score(model_decision_tree, x, y, cv=kf, scoring='neg_mean_absolute_error')
scores_linear_regression_mse = cross_val_score(model_linear_regression, x, y, cv=kf, scoring='neg_mean_squared_error')
scores_linear_regression_mae = cross_val_score(model_linear_regression, x, y, cv=kf, scoring='neg_mean_absolute_error')

avarage_decision_tree_mse_mm = np.mean(scores_decision_tree_mse)
avarage_decision_tree_mae_mm = np.mean(scores_decision_tree_mae)
avarage_linear_regression_mm = np.mean(scores_linear_regression_mse)
avarage_linear_regression_mae_mm = np.mean(scores_linear_regression_mae)

print("-Decision Tree:")
print("MSE: ", avarage_decision_tree_mse_mm)
print("MAE: ", avarage_decision_tree_mae_mm)

print("-Linear Regression:")
print("MSE: ", avarage_linear_regression_mm)
print("MAE: ", avarage_linear_regression_mae_mm)


-Decision Tree:
MSE:  -0.5150365090264244
MAE:  -0.4591538425387597
-Linear Regression:
MSE:  -0.507789387619373
MAE:  -0.5357830764989793


Final Analysis:

In [20]:
analysis = {
    "Model": ["Decision Tree", "Linear Regression"],
    "MSE": [avarage_decision_tree_mse, avarage_linear_regression],
    "MAE": [avarage_decision_tree_mae, avarage_linear_regression_mae],
    "MSE (Power Transform)": [avarage_decision_tree_mse_pt, avarage_linear_regression_pt],
    "MAE (Power Transform)": [avarage_decision_tree_mae_pt, avarage_linear_regression_mae_pt],
    "MSE (Z-scale)": [avarage_decision_tree_mse_z, avarage_linear_regression_z],
    "MAE (Z-scale)": [avarage_decision_tree_mae_z, avarage_linear_regression_mae_z],
    "MSE (Min-max Scale)": [avarage_decision_tree_mse_mm, avarage_linear_regression_mm],
    "MAE (Min-max Scale)": [avarage_decision_tree_mae_mm, avarage_linear_regression_mae_mm]
}

analysis_df = pd.DataFrame(analysis)
analysis_df.to_csv("Analysis.csv", index=False)
print(analysis_df)

               Model       MSE       MAE  MSE (Power Transform)  \
0      Decision Tree -0.511785 -0.462412              -0.519501   
1  Linear Regression -0.530572 -0.531677              -0.569157   

   MAE (Power Transform)  MSE (Z-scale)  MAE (Z-scale)  MSE (Min-max Scale)  \
0              -0.460944      -0.518577      -0.461628            -0.515037   
1              -0.571005      -0.507789      -0.535783            -0.507789   

   MAE (Min-max Scale)  
0            -0.459154  
1            -0.535783  


## Analysis:

- Decision Tree MSE and MAE had small variations using transformation with Min-Max Scaler.
  - Best MSE: -0.511 (Original) ou -0.515 (Min-Max Scale)
  - Best MAE: -0.459 (Min-Max Scale)
- Linear Regression MSE and MAE showed a significant improvement with Power Transformation.
  - Best MSE: -0.507 (Z-Scale e Min-Max Scale)
  - Best MAE: -0.531 (Original) ou -0.535 (Z-Scale ou Min-Max-Scale)

The Decision Tree model did not show any significant impact from transformations, indicating that tree-based models are less sensitive to scaling. On the other hand, Linear Regression demonstrated clear improvements with feature transformations, confirming that linear models are more sensitive to scale and distribution.




